src/traction_modem/Link.hxx

```diff
@@ -0,0 +1,404 @@
+ * modification, are  permitted provided that the following conditions are met:
+ *
+ *  - Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  - Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @file Link.hxx
+ *
+ * Implements link management and acts as a proxy to TxFlow and RxFlow.
+ *
+ * @author Stuart Baker
+ * @date 26 May 2025
+ */
+
+#ifndef _TRACTION_MODEM_LINK_HXX_
+#define _TRACTION_MODEM_LINK_HXX_
+
+#include "traction_modem/RxFlow.hxx"
+#include "traction_modem/TxFlow.hxx"
+
+namespace traction_modem
+{
+
+/// Interface for link start and up/down status.
+class LinkStatusInterface
+{
+protected:
+    /// Called when the link is started.
+    virtual void link_start()
+    {
+    }
+
+    /// Called when link transitions to "up" state.
+    virtual void link_up()
+    {
+    }
+
+    /// Called when link transitions to "down" state.
+    virtual void link_down()
+    {
+    }
+
+    // Protected access for Link object.
+    friend class Link;
+};
+
+/// Link object that holds references to the TX and RX interfaces and tracks
+/// link state.
+class Link : private Atomic
+{
+public:
+    /// Constructor
+    /// @param tx_iface reference to the transmit interface
+    /// @param rx_iface reference to the receive interface
+    Link(TxInterface *tx_iface, RxInterface *rx_iface)
+        : txIface_(tx_iface)
+        , rxIface_(rx_iface)
+        , state_(State::STOP)
+    {
+    }
+
+    /// Check on the link status
+    /// @return true if link is up, else link is down.
+    bool is_link_up()
+    {
+        return state_ == State::UP;
+    }
+
+    /// Bind an interface to the TX and RX interfaces to start transmitting and
+    /// receiving on.
+    /// @param fd interface to transmit and receive messages on
+    void start(int fd)
+    {
+        {
+            AtomicHolder h(this);
+            if (state_ != State::STOP)
+            {
+                // Already started.
```

Comment:
is this a valid sequence of calls? We should at least display an error in the log.